From f197eef036d1d19032638487f7b3ee93df83f8cd Mon Sep 17 00:00:00 2001
From: "Cao,Bing Bu" <bingbu.cao@intel.com>
Date: Wed, 25 Mar 2015 11:33:01 +0000
Subject: [PATCH 327/338] ov5693: Add module parameter "up_delay" for ov5693

Add module parameter up_delay for ov5693

Change-Id: I819f32b2eadfb48352c4bc207f34ddebc3f58a3f
Tracked-On: #JIMINAN-27118
Signed-off-by: Cao,Bing Bu <bingbu.cao@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/47905
Approver: Bin B Han <bin.b.han@intel.com>
Reviewed-by: Bin B Han <bin.b.han@intel.com>
Validation-Android: Wei Tao <wei.tao@intel.com>
Integrator: Wei Tao <wei.tao@intel.com>
Build: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
Maintainer: Wei Tao <wei.tao@intel.com>
---
 .../camera/drivers/media/i2c/ov5693-ecs/ov5693.c   | 24 ++++++++++++++++++----
 1 file changed, 20 insertions(+), 4 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c b/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c
index 8a1a49a..26d3c7f 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c
@@ -41,6 +41,25 @@
 #include "ov5693.h"
 #include "ad5823.h"
 
+#define __cci_delay(t) \
+	do { \
+		if ((t) < 10) { \
+			usleep_range((t) * 1000, ((t) + 1) * 1000); \
+		} else { \
+			msleep((t)); \
+		} \
+	} while (0)
+
+/* Value 30ms reached through experimentation on byt ecs.
+ * The DS specifies a much lower value but when using a smaller value
+ * the I2C bus sometimes locks up permanently when starting the camera.
+ * This issue could not be reproduced on cht, so we can reduce the
+ * delay value to a lower value when insmod.
+ */
+static uint up_delay = 30;
+module_param(up_delay, uint, 0644);
+MODULE_PARM_DESC(up_delay, "Delay prior to the first CCI transaction for ov5693");
+
 static int vcm_ad_i2c_wr8(struct i2c_client *client, u8 reg, u8 val)
 {
 	int err;
@@ -1403,10 +1422,7 @@ static int __power_up(struct v4l2_subdev *sd)
 	if (ret)
 		goto fail_clk;
 
-	/* Value reached through experimentation. The DS specifies a much
-	 * lower value but when using a smaller value the I2C bus sometimes
-	 * locks up permanently when starting the camera. */
-	msleep(30);
+	__cci_delay(up_delay);
 
 	return 0;
 
-- 
1.9.1

